adapters/contxtful/contxtful.go

```diff
@@ -0,0 +1,405 @@
+	MediaType   string              `json:"mediaType"`
+	BidderCode  string              `json:"bidderCode"`
+	PlacementID string              `json:"placementId"`
+	TraceId     string              `json:"traceId,omitempty"`
+	Random      float64             `json:"random,omitempty"`
+	NURL        string              `json:"nurl"`
+	BURL        string              `json:"burl"`
+	LURL        string              `json:"lurl,omitempty"`
+	Ext         jsonutil.RawMessage `json:"ext,omitempty"`
+}
+
+type BidProcessingContext struct {
+	request        *openrtb2.BidRequest
+	requestData    *adapters.RequestData
+	customerId     string
+	bidderResponse *adapters.BidderResponse
+	errors         []error
+	// Event URL generation fields
+	version     string
+	domain      string
+	adRequestID string
+}
+
+type adapter struct {
+	endpointTemplate *template.Template
+}
+
+// Essential constants
+const (
+	BidderName     = "contxtful"
+	DefaultVersion = "v1"
+	FieldBidder    = "bidder"
+	PbsPath        = "/pbs/"
+)
+
+// Helper to safely extract bidder parameters from impression extension
+func extractBidderParams(impExt jsonutil.RawMessage) (*openrtb_ext.ExtImpContxtful, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := jsonutil.Unmarshal(impExt, &bidderExt); err != nil {
+		return nil, err
+	}
+
+	var contxtfulParams openrtb_ext.ExtImpContxtful
+	if err := jsonutil.Unmarshal(bidderExt.Bidder, &contxtfulParams); err != nil {
+		return nil, err
+	}
+
+	return &contxtfulParams, nil
+}
+
+// Builder builds a new instance of the Contxtful adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	return &adapter{
+		endpointTemplate: template.Must(template.New("endpointTemplate").Parse(config.Endpoint)),
+	}, nil
+}
+
+// Extract impression parameters (validation handled by JSON schema)
+func validateImpressions(request *openrtb2.BidRequest) ([]string, string, []error) {
+	var errors []error
+	var customerId string
+	validPlacements := make([]string, 0, len(request.Imp))
+
+	for _, imp := range request.Imp {
+		contxtfulParams, err := extractBidderParams(imp.Ext)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		if customerId == "" {
+			customerId = contxtfulParams.CustomerId
+		}
+		validPlacements = append(validPlacements, contxtfulParams.PlacementId)
+	}
+
+	return validPlacements, customerId, errors
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errors []error
+
+	// Build headers efficiently
+	headers := http.Header{
+		"Content-Type": []string{"application/json;charset=utf-8"},
+		"Accept":       []string{"application/json"},
+	}
+
+	// Add privacy-related headers if available
+	if reqInfo != nil && reqInfo.GlobalPrivacyControlHeader != "" {
+		headers.Set("Sec-GPC", reqInfo.GlobalPrivacyControlHeader)
```

Comment:
Update an existing or add a new JSON test to cover this path.

Reply:
Hi,
It seems that Prebid Server core already does that freely here: https://github.com/prebid/prebid-server/blob/v3.25.0/exchange/bidder.go#L223

So I removed this code.
see commit 64d5b4263c15f96070bfa988d3800b524a35aff2